lib/charms/saml_integrator/v0/saml.py

```diff
@@ -0,0 +1,253 @@
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+# pylint: disable=wrong-import-position
+import re
+from typing import Dict, List, Optional
+
+import ops
+from pydantic import AnyHttpUrl, BaseModel, Field
+from pydantic.tools import parse_obj_as
+
+DEFAULT_RELATION_NAME = "saml"
+
+
+class SamlEndpoint(BaseModel):
+    """Represent a SAML endpoint.
+
+    Attrs:
+        name: Endpoint name.
+        url: Endpoint URL.
+        binding: Endpoint binding.
+        response_url: URL to address the response to.
+    """
+
+    name: str = Field(..., min_length=1)
+    url: AnyHttpUrl
+    binding: str = Field(..., min_length=1)
+    response_url: Optional[AnyHttpUrl]
+
+    def to_relation_data(self) -> Dict[str, str]:
+        """Convert an instance of SamlEndpoint to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        result: Dict[str, str] = {}
+        # Get the HTTP method from the SAML binding
+        http_method = self.binding.split(":")[-1].split("-")[-1].lower()
+        # Transform name into snakecase
+        lowercase_name = re.sub(r"(?<!^)(?=[A-Z])", "_", self.name).lower()
+        prefix = f"{lowercase_name}_{http_method}_"
+        result[f"{prefix}url"] = self.url
+        result[f"{prefix}binding"] = self.binding
+        if self.response_url:
+            result[f"{prefix}response_url"] = self.response_url
+        return result
+
+    @classmethod
+    def from_relation_data(cls, relation_data: Dict[str, str]) -> "SamlEndpoint":
+        """Initialize a new instance of the SamlEndpoint class from the relation data.
+
+        Args:
+            relation_data: the relation data.
+
+        Returns:
+            A SamlEndpoint instance.
+        """
+        url_key = ""
+        for key in relation_data:
+            if key.endswith("_redirect_url") or key.endswith("_post_url"):
+                url_key = key
+        lowercase_name = "_".join(url_key.split("_")[:-2])
+        name = "".join(x.capitalize() for x in lowercase_name.split("_"))
+        http_method = url_key.split("_")[-2]
+        prefix = f"{lowercase_name}_{http_method}_"
+        return cls(
+            name=name,
+            url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}url"]),
+            binding=relation_data[f"{prefix}binding"],
+            response_url=parse_obj_as(AnyHttpUrl, relation_data[f"{prefix}response_url"])
+            if f"{prefix}response_url" in relation_data
+            else None,
+        )
+
+
+class SamlRelationData(BaseModel):
+    """Represent the relation data.
+
+    Attrs:
+        entity_id: SAML entity ID.
+        metadata_url: URL to the metadata.
+        certificates: List of SAML certificates.
+        endpoints: List of SAML endpoints.
+    """
+
+    entity_id: str = Field(..., min_length=1)
+    metadata_url: AnyHttpUrl
+    certificates: List[str]
+    endpoints: List[SamlEndpoint]
```

Comment:
could this be a tuple? And can the built in `list` be used?

Reply:
Indeed. Changed. Thx